File: src/main/java/eu/mshade/enderman/metadata/itemstack/LoreItemStackMetadataWrapper.java
package eu.mshade.enderman.metadata.itemstack;

import eu.mshade.enderframe.item.ItemStack;
import eu.mshade.enderframe.item.ItemStackMetadataWrapper;
import eu.mshade.enderframe.item.metadata.LoreItemStackMetadata;
import eu.mshade.enderframe.metadata.MetadataKeyValueBucket;
import eu.mshade.enderframe.item.ItemStackMetadataKey;
import eu.mshade.enderframe.mojang.chat.TextComponent;
import eu.mshade.mwork.binarytag.BinaryTagType;
import eu.mshade.mwork.binarytag.StringBinaryTag;
import eu.mshade.mwork.binarytag.entity.CompoundBinaryTag;
import eu.mshade.mwork.binarytag.entity.ListBinaryTag;

import java.util.ArrayList;
import java.util.List;

public class LoreItemStackMetadataWrapper implements ItemStackMetadataWrapper {

    @Override
    public void write(CompoundBinaryTag compoundBinaryTag, ItemStack itemStack) {
        MetadataKeyValueBucket metadataKeyValueBucket = itemStack.getMetadataKeyValueBucket();
        LoreItemStackMetadata loreItemStackMetadata = metadataKeyValueBucket.getMetadataKeyValue(ItemStackMetadataKey.LORE, LoreItemStackMetadata.class);
        CompoundBinaryTag displayCompoundBinaryTag = compoundBinaryTag.computeIfAbsent("display", s -> new CompoundBinaryTag());
        List<TextComponent> lore = loreItemStackMetadata.getMetadataValue();
        if (!lore.isEmpty()) {
            ListBinaryTag listBinaryTag = new ListBinaryTag(BinaryTagType.STRING);
            lore.forEach(s -> listBinaryTag.add(new StringBinaryTag(s.toLegacyText())));
            displayCompoundBinaryTag.putBinaryTag("Lore", listBinaryTag);
        }
    }

    @Override
    public void read(CompoundBinaryTag compoundBinaryTag, ItemStack itemStack) {
        MetadataKeyValueBucket metadataKeyValueBucket = itemStack.getMetadataKeyValueBucket();
        if (!compoundBinaryTag.containsKey("display")) return;
        CompoundBinaryTag display = (CompoundBinaryTag) compoundBinaryTag.getBinaryTag("display");
        if (!display.containsKey("Lore")) return;
        ListBinaryTag lore = (ListBinaryTag) display.getBinaryTag("Lore");
        LoreItemStackMetadata loreItemStackMetadata = new LoreItemStackMetadata(new ArrayList<>());
        lore.getValue().forEach(binaryTag -> loreItemStackMetadata.getMetadataValue().add(TextComponent.of((String) binaryTag.getValue())));
        metadataKeyValueBucket.setMetadataKeyValue(loreItemStackMetadata);
    }
}

File: src/main/java/eu/mshade/enderman/metadata/itemstack/LoreItemStackMetadataWrapper.kt
package eu.mshade.enderman.metadata.itemstack

import eu.mshade.enderframe.item.ItemStack
import eu.mshade.enderframe.item.ItemStackMetadataKey
import eu.mshade.enderframe.item.ItemStackMetadataWrapper
import eu.mshade.enderframe.item.metadata.LoreItemStackMetadata
import eu.mshade.enderframe.mojang.chat.TextComponent
import eu.mshade.mwork.binarytag.BinaryTag
import eu.mshade.mwork.binarytag.BinaryTagType
import eu.mshade.mwork.binarytag.StringBinaryTag
import eu.mshade.mwork.binarytag.entity.CompoundBinaryTag
import eu.mshade.mwork.binarytag.entity.ListBinaryTag
import java.util.function.Consumer

class LoreItemStackMetadataWrapper : ItemStackMetadataWrapper {
    override fun write(compoundBinaryTag: CompoundBinaryTag, itemStack: ItemStack) {
        val metadataKeyValueBucket = itemStack.metadatas
        val loreItemStackMetadata =
            metadataKeyValueBucket.getMetadataKeyValue(ItemStackMetadataKey.LORE) as LoreItemStackMetadata
        val displayCompoundBinaryTag =
            compoundBinaryTag.computeIfAbsent("display") { _ -> CompoundBinaryTag() }
        val lore = loreItemStackMetadata.metadataValue
        if (!lore.isEmpty()) {
            val listBinaryTag = ListBinaryTag(BinaryTagType.STRING)
            lore.forEach(Consumer { s: TextComponent -> listBinaryTag.add(StringBinaryTag(s.toLegacyText())) })
            displayCompoundBinaryTag.putBinaryTag("Lore", listBinaryTag)
        }
    }

    override fun read(compoundBinaryTag: CompoundBinaryTag, itemStack: ItemStack) {
        val metadataKeyValueBucket = itemStack.metadatas
        if (!compoundBinaryTag.containsKey("display")) return
        val display = compoundBinaryTag.getBinaryTag("display") as CompoundBinaryTag?
        if (!display!!.containsKey("Lore")) return
        val lore = display.getBinaryTag("Lore") as ListBinaryTag?
        val loreItemStackMetadata = LoreItemStackMetadata(ArrayList())
        lore!!.value.forEach(Consumer<BinaryTag<*>> { binaryTag: BinaryTag<*> ->
            loreItemStackMetadata.metadataValue.add(
                TextComponent.of(binaryTag.value as String?)
            )
        })
        metadataKeyValueBucket.setMetadataKeyValue(loreItemStackMetadata)
    }
}
